node/core/prospective-parachains/src/tests.rs

```diff
@@ -0,0 +1,1249 @@
+	assert_eq!(view.candidate_storage.get(&1.into()).unwrap().len(), (2, 2));
+	assert_eq!(view.candidate_storage.get(&2.into()).unwrap().len(), (0, 0));
+}
+
+// Test depth and pvd queries.
+#[test]
+fn check_depth_and_pvd_queries() {
+	let test_state = TestState::default();
+	let view = test_harness(|mut virtual_overseer| async move {
+		// Leaf A
+		let leaf_a = TestLeaf {
+			number: 100,
+			hash: Hash::from_low_u64_be(130),
+			para_data: vec![
+				(1.into(), PerParaData::new(97, HeadData(vec![1, 2, 3]))),
+				(2.into(), PerParaData::new(100, HeadData(vec![2, 3, 4]))),
+			],
+		};
+
+		// Activate leaves.
+		activate_leaf(&mut virtual_overseer, &leaf_a, &test_state).await;
+
+		// Candidate A.
+		let (candidate_a, pvd_a) = make_candidate(
+			&leaf_a,
+			1.into(),
+			HeadData(vec![1, 2, 3]),
+			HeadData(vec![1]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_a = candidate_a.hash();
+		let response_a = vec![(leaf_a.hash, vec![0])];
+
+		// Candidate B.
+		let (candidate_b, pvd_b) = make_candidate(
+			&leaf_a,
+			1.into(),
+			HeadData(vec![1]),
+			HeadData(vec![2]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_b = candidate_b.hash();
+		let response_b = vec![(leaf_a.hash, vec![1])];
+
+		// Candidate C.
+		let (candidate_c, pvd_c) = make_candidate(
+			&leaf_a,
+			1.into(),
+			HeadData(vec![2]),
+			HeadData(vec![3]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_c = candidate_c.hash();
+		let response_c = vec![(leaf_a.hash, vec![2])];
+
+		// Get hypothetical depth and pvd of candidate A before adding it.
+		get_hypothetical_depth(
+			&mut virtual_overseer,
+			candidate_hash_a,
+			1.into(),
+			leaf_a.para_data(1.into()).head_data.clone(),
+			leaf_a.hash,
+			leaf_a.hash,
+			vec![0],
+		)
+		.await;
+		get_pvd(
+			&mut virtual_overseer,
+			1.into(),
+			leaf_a.hash,
+			leaf_a.para_data(1.into()).head_data.clone(),
+			Some(pvd_a.clone()),
+		)
+		.await;
+
+		// Add candidate A.
+		second_candidate(
+			&mut virtual_overseer,
+			candidate_a.clone(),
+			pvd_a.clone(),
+			response_a.clone(),
+		)
+		.await;
+
+		// Get depth and pvd of candidate A after adding it.
+		get_hypothetical_depth(
+			&mut virtual_overseer,
+			candidate_hash_a,
+			1.into(),
+			HeadData(vec![1, 2, 3]),
+			// TODO: When would these be different?
+			leaf_a.hash,
+			leaf_a.hash,
+			vec![0],
+		)
+		.await;
+		get_pvd(
+			&mut virtual_overseer,
+			1.into(),
+			leaf_a.hash,
+			HeadData(vec![1, 2, 3]),
+			Some(pvd_a.clone()),
+		)
+		.await;
+
+		// TODO: Get hypothetical depth and pvd of candidate B before adding it.
+		get_hypothetical_depth(
+			&mut virtual_overseer,
+			candidate_hash_b,
+			1.into(),
+			HeadData(vec![1]),
+			leaf_a.hash,
+			leaf_a.hash,
+			vec![1],
+		)
+		.await;
+		// get_pvd(
+		// 	&mut virtual_overseer,
+		// 	1.into(),
+		// 	leaf_a.hash,
+		// 	HeadData(vec![1]),
+		// 	Some(pvd_b.clone()),
+		// )
+		// .await;
```

Comment:
Is `GetProspectiveValidationData` supposed to be able to return PVD for a candidate before it's been added? I ask because of this:

```rs
/// Get the validation data of some prospective candidate. The candidate doesn't need
/// to be part of any fragment tree, but this only succeeds if the parent head-data and
/// relay-parent are part of some fragment tree.
```

Here I'm querying by the head-data of the parent candidate, but nothing is returned. Wanted to make sure I'm not misunderstanding before diving more deeply.

Reply:
This request is not related to any particular candidate, note that it doesn't reference candidate hash.
```rust
/// A request for the persisted validation data stored in the prospective
/// parachains subsystem.
#[derive(Debug)]
pub struct ProspectiveValidationDataRequest {
	/// The para-id of the candidate.
	pub para_id: ParaId,
	/// The relay-parent of the candidate.
	pub candidate_relay_parent: Hash,
	/// The parent head-data hash.
	pub parent_head_data_hash: Hash,
}
```
If a collator advertises some candidate, a validator asks prospective parachains to construct PVD for it, if possible.

Reply:
As for the test itself, of course it wouldn't find PVD for a parent head and para which are not present in any fragment tree

Reply:
Shouldn't the parent head be present in the fragment tree? I don't understand why it's not. Is there a way to fix the test?

Reply:
> Shouldn't the parent head be present in the fragment tree? I don't understand why it's not. Is there a way to fix the test?

Not necessarily. The parent head might be the "root" of the fragment tree, i.e. it is the `HeadData` stored in the relay-chain state.

> Is GetProspectiveValidationData supposed to be able to return PVD for a candidate before it's been added? I ask because of this:

The answer is "yes", as @slumber commented. We can predict the `PersistedValidationData` for any _possible child_ of either the root or a candidate within the fragment tree, and in fact we need to do this when seconding candidates, and on collators generating candidates.

Reply:
@slumber 

`add_candiddate` looks fine to me so far.

The index is `by_parent_head` and maps parent head hashes to candidates building upon that parent head. Using the para head hash the candidate produces would be a mapping `by_output_head`. 

We use `by_parent_head` to search for candidates building on top of output heads when populating the fragment tree in `add_and_populate`, as we are looking for _child_ candidates.

In `head_data_by_hash`, we are looking for `HeadData` that has a specific hash H. We have a two-part algorithm:
  1. Find a candidate which builds upon `HeadData` with hash H by searching `by_parent_head`, i.e. find a child candidate of H
  2. Copy the `parent_head` from the persisted validation data of that child block.

This algorithm looks correct to me, and I believe that tests should confirm it.

Reply:
Tests confirm that we have an issue in `answer_prospective_validation_data_request`.

Suppose committed `para_head` is `A`. We introduce `Candidate_A_B` which builds on top of para head `A` and produces `B`.
When we attempt to fetch PVD for `Candidate_B_C` it will provide `B` as parent_head_data.

```rust
let mut head_data =
		storage.head_data_by_hash(&request.parent_head_data_hash).map(|x| x.clone());
```

Obviously there's no such entry and consequently no candidate will be accepted by collator protocol due to missing pvd (except for 0-depth one)

Reply:
Aha. I am confirming locally as well; as you describe it's a legitimate issue, although the `by_parent_head` mapping is not the issue. `head_data_by_hash` should use a different algorithm that also can get output head data of stored candidates, perhaps with introducing a new mapping.

Reply:
https://github.com/paritytech/polkadot/commit/a32f0cd6ad46f17bfee653d75f9d79727d4cd7a6  <- this commit fixes the issue correctly. needs a new test for candidate storage pruning to make sure the new index is cleaned up correctly.

Reply:
@rphmeier Done! The test passes now.

> needs a new test for candidate storage pruning to make sure the new index is cleaned up correctly.

Should I do that on this branch?

Reply:
yes

Reply:
Done. 🙂